### PR TITLE
[vcpkg/toolchain] set CMAKE_SYSTEM_PROCESSOR for arm on windows

### LINKED ISSUE #9390: Missing CMAKE_SYSTEM_PROCESSOR variable in cross-compiling
`CMAKE_SYSTEM_PROCESSOR` is not set when building an ARM / ARM64 target, so configure does not work correctly with some libraries. (e.g opencv)

For ARM64 Windows 10 (not UWP), `CMAKE_SYSTEM_PROCESSOR` and `CMAKE_CROSSCOMPILING` is not set.

**Proposed solution**

Set `CMAKE_SYSTEM_PROCESSOR` and `CMAKE_CROSSCOMPILING` when building ARM / ARM64 target using toolchain file.

I created a branch to check the operation.

- Explicitly set `CMAKE_SYSTEM_PROCESSOR` only for ARM / ARM64
- For ARM64 Windows 10 (not UWP), set `CMAKE_SYSTEM_NAME` and `CMAKE_SYSTEM_VERSION`
  - (to set `CMAKE_SYSTEM_PROCESSOR` and `CMAKE_CROSSCOMPILING`)
- Turn off `CMAKE_CROSSCOMPILING` for x86 / x64 UWP

https://github.com/shibayan/vcpkg/tree/crosscompile

**Additional context**

Share the OpenCV 4 configure log. In the modified toolchains, the CPU is correctly detected.

command : `vcpkg install opencv4[core]:arm64-uwp`

**Configure log with master branch**

```
-- Detecting C compile features
-- Detecting C compile features - done
-- Detected processor: 
-- Performing Test HAVE_CXX11 (check file: cmake/checks/cxx11.cpp)
-- Performing Test HAVE_CXX11 - Success
```

**Configure log with crosscompile branch**

```
-- Detecting C compile features
-- Detecting C compile features - done
-- Detected processor: ARM64
-- Performing Test HAVE_CXX11 (check file: cmake/checks/cxx11.cpp)
-- Performing Test HAVE_CXX11 - Success
```

Rel #3036 #4907

### PR DESCRIPTION
**Pull request by Neumann-A**
Correct the windows toolchain to correctly set CMAKE_SYSTEM_PROCESSOR so that CMake goes into CROSSCOMPILING mode. 

closes #9390

### CODE REVIEW

**Comment by cenit** [comment]
as a side note, opencv is already compatible with arm64/arm cross compiling without this PR in any case

**Comment by Neumann-A** [comment]
@cenit Did you patch it to mark it work or does it work by default? I interpreted #9390 as an request to correct the toolchain and obviously the toolchain is currently wrong for arm builds. 

@PhoebeHui : I am not going to correct the baseline arm errors here. I just want to make it obvious that the arm setup is currently incorrect from a CMake based point of view.

**Comment by PhoebeHui** [comment]
@Neumann-A, I will take a look at the baseline issues.

**Comment by cenit** [comment]
There are some specific lines in the portfile to deal with  the issue: https://github.com/microsoft/vcpkg/blob/master/ports/opencv4/portfile.cmake#L27-L35

File: scripts/toolchains/windows.cmake [hunk 1]
```diff
@@ -2,6 +2,14 @@ if(NOT _VCPKG_WINDOWS_TOOLCHAIN)
 set(_VCPKG_WINDOWS_TOOLCHAIN 1)
 set(CMAKE_MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>$<$<STREQUAL:${VCPKG_CRT_LINKAGE},dynamic>:DLL>" CACHE STRING "")
 
+if(VCPKG_TARGET_ARCHITECTURE STREQUAL "arm")
+    set(CMAKE_SYSTEM_NAME "${CMAKE_HOST_SYSTEM_NAME}" CACHE STRING "") # forces CMAKE_CROSSCOMPILING
+    set(CMAKE_SYSTEM_PROCESSOR armv7l CACHE STRING "")
+elseif(VCPKG_TARGET_ARCHITECTURE STREQUAL "arm64")
+    set(CMAKE_SYSTEM_NAME "${CMAKE_HOST_SYSTEM_NAME}" CACHE STRING "") # forces CMAKE_CROSSCOMPILING
+    set(CMAKE_SYSTEM_PROCESSOR aarch64 CACHE STRING "")
```
Comment (ras0219-msft) [inline scripts/toolchains/windows.cmake:10]: Based on https://github.com/Kitware/CMake/blob/22902cb38cdc934cb113269ef89ee30fba11391c/Modules/CMakeDetermineSystem.cmake#L105-L112
```cmake
if(CMAKE_HOST_WIN32)
 if (DEFINED ENV{PROCESSOR_ARCHITEW6432})
  set (CMAKE_HOST_SYSTEM_PROCESSOR "$ENV{PROCESSOR_ARCHITEW6432}")
 else()
  set (CMAKE_HOST_SYSTEM_PROCESSOR "$ENV{PROCESSOR_ARCHITECTURE}")
 endif()
endif()
```
and https://docs.microsoft.com/en-us/windows/win32/winprog64/wow64-implementation-details#environment-variables
```
(64-bit) PROCESSOR_ARCHITECTURE=AMD64 or PROCESSOR_ARCHITECTURE=IA64 or PROCESSOR_ARCHITECTURE=ARM64
(32-bit) PROCESSOR_ARCHITECTURE=x86 PROCESSOR_ARCHITEW6432=%PROCESSOR_ARCHITECTURE%
```

I think these should be `ARM` and `ARM64` respectively.

File: scripts/toolchains/windows.cmake [hunk 2]
```diff
@@ -2,6 +2,14 @@ if(NOT _VCPKG_WINDOWS_TOOLCHAIN)
 set(_VCPKG_WINDOWS_TOOLCHAIN 1)
 set(CMAKE_MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>$<$<STREQUAL:${VCPKG_CRT_LINKAGE},dynamic>:DLL>" CACHE STRING "")
 
+if(VCPKG_TARGET_ARCHITECTURE STREQUAL "arm")
+    set(CMAKE_SYSTEM_NAME "${CMAKE_HOST_SYSTEM_NAME}" CACHE STRING "") # forces CMAKE_CROSSCOMPILING
+    set(CMAKE_SYSTEM_PROCESSOR armv7l CACHE STRING "")
+elseif(VCPKG_TARGET_ARCHITECTURE STREQUAL "arm64")
+    set(CMAKE_SYSTEM_NAME "${CMAKE_HOST_SYSTEM_NAME}" CACHE STRING "") # forces CMAKE_CROSSCOMPILING
+    set(CMAKE_SYSTEM_PROCESSOR aarch64 CACHE STRING "")
+endif()
```
Comment (ras0219-msft) [inline scripts/toolchains/windows.cmake:11]: We should really add x86 and x64 as well.

**Review by ras0219-msft — CHANGES_REQUESTED** [review]
This is an important fix, thank you!

File: scripts/toolchains/windows.cmake (hunk 1 above)
Comment (Neumann-A) [reply scripts/toolchains/windows.cmake:10]: > I think these should be ARM and ARM64 respectively.

I have doubts about that. Probably no port will be "correctly" checking against "purely MS defined" architectures here.

Comment (ras0219) [reply scripts/toolchains/windows.cmake:10]: That may be true, but these are the official values and must be this way to work against `if(CMAKE_SYSTEM_PROCESSOR STREQUAL CMAKE_HOST_SYSTEM_PROCESSOR)`

Comment (davidebeatrici) [reply scripts/toolchains/windows.cmake:10]: I confirm `arm64` is the correct value: https://github.com/mumble-voip/mumble/pull/4765/commits/2996e9f72ae3a8ccc5a9bdf6bd44eaea7c6b3425

**Comment by dscho** [comment]
@Neumann-A gentle ping: was there any progress?

**Comment by PhoebeHui** [comment]
CI failures:
- [X] forest: known baseline issue, would be fixed by PR https://github.com/microsoft/vcpkg/pull/16836
- [ ] vc:x64-uwp
- [ ] kfr:x64-uwp
- [ ] ampl-mp:arm64-windows
- [ ] corrade:arm64-windows
- [ ] gainput:arm64-windows
- [ ] glog:arm64-windows
- [ ] gppanel:arm64-windows
- [ ] orc:arm64-windows
- [X] fltk:arm64-windows
- [ ] halide:x86-windows
- [ ] dcmtk:x86-windows

**Comment by Neumann-A** [comment]
Qt6 and the cmake test port require these changes to correctly compile for ARM

File: scripts/toolchains/windows.cmake [hunk 3]
```diff
@@ -2,6 +2,18 @@ if(NOT _VCPKG_WINDOWS_TOOLCHAIN)
 set(_VCPKG_WINDOWS_TOOLCHAIN 1)
 set(CMAKE_MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>$<$<STREQUAL:${VCPKG_CRT_LINKAGE},dynamic>:DLL>" CACHE STRING "")
 
+if(VCPKG_TARGET_ARCHITECTURE STREQUAL "x86")
+    set(CMAKE_SYSTEM_PROCESSOR x86 CACHE STRING "")
+elseif(VCPKG_TARGET_ARCHITECTURE STREQUAL "x64")
+    set(CMAKE_SYSTEM_PROCESSOR AMD64 CACHE STRING "")
+elseif(VCPKG_TARGET_ARCHITECTURE STREQUAL "arm")
+    set(CMAKE_SYSTEM_NAME "${CMAKE_HOST_SYSTEM_NAME}" CACHE STRING "") # forces CMAKE_CROSSCOMPILING
```
Comment (strega-nil-ms) [inline scripts/toolchains/windows.cmake:10]: what if you're building on ARM64?

**Comment by JackBoosY** [comment]
Ping for response.

**Comment by dg0yt** [comment]
Note that the `windows.cmake toolchain` file is not loaded by `vcpkg_cmake_configure` at the moment, https://github.com/microsoft/vcpkg/pull/21857.

> This is an important fix, thank you!

And it still is, 10 months later. Some packages rely on CMAKE_SYSTEM_PROCESSOR being defined.(`if(${CMAKE_SYSTEM_PROCESSOR} STREQUAL "x86")` causes a cmake error.)

**Comment by JackBoosY** [comment]
Any progress here?

**Comment by Neumann-A** [comment]
superseded by https://github.com/microsoft/vcpkg/pull/24100

**Pull request closed without merging**